### PR TITLE
[AI-100]: fix intermittent Perplexity 400 and flatten log newlines

### PR DESCRIPTION
**Pull request by stephenfuqua**
## Summary

- **Bug fix:** `truncateToCharBudget` in `thread-history.js` could expose a leading `assistant` message after trimming the conversation history, causing Perplexity to return a `400 — messages must alternate` error. Re-applying the leading-assistant-drop guard after truncation prevents this.
- **Logging:** Replaced Bolt's default `ConsoleLogger` with a custom single-line logger that replaces literal newlines (`\n`) with the two-character sequence `\n` in all log output, preventing interleaved log entries in Azure's log stream when concurrent async operations log multi-line Error stack traces.

Closes AI-100.

## Root Cause

`normalizeMessages` strips leading `assistant` messages, but `truncateToCharBudget` runs afterward and removes from the front of the array — which can re-expose an `assistant` message at position 0. This only triggers when a conversation exceeds 20,000 chars and the cut lands on an odd message boundary, explaining the intermittent nature.

## Test Plan

- [ ] New regression test in `tests/agent/thread-history.test.js` covers the truncation edge case (was failing before the fix, passes after).
- [ ] Full test suite: `npm test` — 227 tests passing.
- [ ] Deploy to staging and confirm the Perplexity 400 error no longer occurs in long conversations.
- [ ] Confirm Azure log output shows single-line entries (no interleaving) for concurrent requests.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/fiona-slack/src/agent/thread-history.js [hunk 1]
```diff
@@ -118,6 +118,13 @@ export async function buildThreadHistory(
   const normalized = normalizeMessages(rawMessages);
   const trimmed = truncateToCharBudget(normalized, maxChars);
 
+  // Truncation removes from the front, which can expose a leading assistant message
+  // if the oldest user message was the last one removed. Re-apply the same guard as
+  // normalizeMessages so the first message is always from the user.
+  while (trimmed.length > 0 && trimmed[0].role === 'assistant') {
+    trimmed.shift();
+  }
```
[the remark this answers is not in the document]
Comment (stephenfuqua) [reply apps/fiona-slack/src/agent/thread-history.js]: @copilot apply changes based on [this feedback](https://github.com/Ed-Fi-Alliance-OSS/Fiona/pull/32#discussion_r3094976160)

File: apps/fiona-slack/src/agent/thread-history.js [hunk 2]
```diff
@@ -119,11 +128,9 @@ export async function buildThreadHistory(
...
 
   // Truncation removes from the front, which can expose a leading assistant message
-  // if the oldest user message was the last one removed. Re-apply the same guard as
-  // normalizeMessages so the first message is always from the user.
-  while (trimmed.length > 0 && trimmed[0].role === 'assistant') {
-    trimmed.shift();
-  }
+  // if the oldest user message was the last one removed. Re-apply the guard so the
+  // first message is always from the user.
+  dropLeadingAssistantMessages(trimmed);
```
Comment (stephenfuqua) [inline apps/fiona-slack/src/agent/thread-history.js]: @copilot shouldn't this assigned the return from `dropLeadingAssistantMessages` back to `trimmed`?

File: apps/fiona-slack/src/agent/thread-history.js [hunk 3]
```diff
@@ -12,6 +12,21 @@ const MAX_HISTORY_CHARS = 20_000;
...
+ * Removes leading assistant messages from a message array in-place.
+ * LLM APIs require the first non-system message to be from the user; injecting a
+ * synthetic user placeholder would pollute the context with content the user never sent.
+ *
+ * @param {Array<{role: string, content: string}>} messages - Mutated in place.
+ * @returns {Array<{role: string, content: string}>} The same array reference.
+ */
+function dropLeadingAssistantMessages(messages) {
+  while (messages.length > 0 && messages[0].role === 'assistant') {
+    messages.shift();
```
Comment (roberthunterjr) [inline apps/fiona-slack/src/agent/thread-history.js:25]:
```suggestion
  const result = [...messages];
  while (result.length > 0 && result[0].role === 'assistant') {
    result.shift();
```

File: apps/fiona-slack/src/agent/thread-history.js [hunk 4]
```diff
@@ -12,6 +12,21 @@ const MAX_HISTORY_CHARS = 20_000;
...
+  }
+  return messages;
```
Comment (roberthunterjr) [inline apps/fiona-slack/src/agent/thread-history.js:27]:
```suggestion
  return result;
```

File: apps/fiona-slack/src/agent/thread-history.js [hunk 5]
```diff
@@ -12,6 +12,21 @@ const MAX_HISTORY_CHARS = 20_000;
...
+}
```
Comment (roberthunterjr) [inline apps/fiona-slack/src/agent/thread-history.js:28]: There was an interesting point that came up from an in-line review but didn't for the overall PR review. It mentioned that `messages` here is being mutated directly without making a copy, which is true.

This is different from the `truncateToCharBudget()` function on L64 that creates a copy via `const result =[...messages]` array expansion and returns that as the modified array, leaving the original in tact. Creating a suggestion manually for you to review.

File: apps/fiona-slack/src/agent/thread-history.js (hunk 3 above)
Comment (stephenfuqua) [reply apps/fiona-slack/src/agent/thread-history.js:25]: Why treat `messages` as immutable? I'm going to look more deeply at the code to try to answer this on my own. Leaving the comment so that I don't forget to look in the morning @stephenfuqua

Comment (stephenfuqua) [reply apps/fiona-slack/src/agent/thread-history.js:25]: This is an interesting nitpick, but not worthwhile and eventually slightly detrimental.

There is no need to preserve the original array. The code was originally inline and modification made perfect sense. A review comment caused us to notice that there is duplication, so it was moved out to a function.

I added a return statement to the function, returning the original array, because it feels slightly more transparent to write `a = function(a)` than just `function(a)` insofar as the former clearly signals that something has changed, and it is ambiguous in the latter.

File: apps/fiona-slack/src/agent/thread-history.js (hunk 4 above)
Comment (stephenfuqua) [reply apps/fiona-slack/src/agent/thread-history.js:27]: Won't do.

File: apps/fiona-slack/src/agent/thread-history.js (hunk 5 above)
Comment (stephenfuqua) [reply apps/fiona-slack/src/agent/thread-history.js:28]: Won't do.

There _is_ a purity to treating input as immutable. But there's also a cost in that we now have an unnecessary copy of the array. If the array is very large, now we have two things to garbage collect.

When I reviewed `truncateToCharBudget` either I did not pay attention to the array copy, or I passed over it as too insignificant for comment.

**Review by roberthunterjr — APPROVED** [review]
